## **UAS Data Wranling**
### Data Cleaning Transjakarta

In [1]:
# Library import
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

Keterangan dan Informasi data :

- transID: ID unik transaksi untuk setiap transaksi.
- payCardID: Identifikasi utama pelanggan. Kartu yang digunakan pelanggan sebagai tiket masuk dan keluar.
- payCardBank: Nama penerbit bank kartu pelanggan.
- payCardName: Nama pelanggan yang terdapat di dalam kartu.
- payCardSex: Jenis kelamin pelanggan yang terdapat di dalam kartu.
- payCardBirthDate: Tahun kelahiran pelanggan.
- corridorID: ID Koridor / ID Rute sebagai kunci untuk pengelompokan rute.
- corridorName: Nama Koridor / Nama Rute yang berisi titik awal dan akhir untuk setiap rute.
- direction: 0 untuk Pergi, 1 untuk Kembali. Arah rute.
- tapInStops: ID Titik Tap In (masuk) untuk mengidentifikasi nama halte.
- tapInStopsName: Nama Halte Titik Tap In di mana pelanggan melakukan tap masuk.
- tapInStopsLat: Garis lintang Halte Titik Tap In.
- tapInStopsLon: Garis bujur Halte Titik Tap In.
- stopStartSeq: Urutan berhenti, mulai dari yang pertama, kedua, dst. Terkait dengan arah.
- tapInTime: Waktu tap masuk. Tanggal dan waktu.
- tapOutStops: ID Titik Tap Out (keluar) untuk mengidentifikasi nama halte.
- tapOutStopsName: Nama Halte Titik Tap Out di mana pelanggan melakukan tap keluar.
- tapOutStopsLat: Garis lintang Halte Titik Tap Out.
- tapOutStopsLon: Garis bujur Halte Titik Tap Out.
- stopEndSeq: Urutan berhenti, mulai dari yang pertama, kedua, dst. Terkait dengan arah.
- tapOutTime: Waktu tap keluar. Tanggal dan waktu.
- payAmount: Jumlah yang dibayar oleh pelanggan. Beberapa gratis. Beberapa tidak.

In [2]:
# Load the dataset
df = pd.read_csv('Transjakarta.csv')
display(df.head(3), df.tail(3))

transID         payCardID payCardBank     payCardName payCardSex  \
0  EIIW227B8L34VB   180062659848800      emoney  Bajragin Usada          M   
1  LGXO740D2N47GZ  4885331907664776         dki    Gandi Widodo          F   
2  DJWR385V2U57TO  4996225095064169         dki   Emong Wastuti          F   

   payCardBirthDate corridorID                              corridorName  \
0              2008          5                     Matraman Baru - Ancol   
1              1997         6C  Stasiun Tebet - Karet via Patra Kuningan   
2              1992        R1A                        Pantai Maju - Kota   

   direction tapInStops  ... tapInStopsLon  stopStartSeq            tapInTime  \
0        1.0     P00142  ...     106.84402             7  2023-04-03 05:21:44   
1        0.0    B01963P  ...     106.83302            13  2023-04-03 05:42:44   
2        0.0    B00499P  ...     106.81435            38  2023-04-03 05:59:06   

   tapOutStops        tapOutStopsName tapOutStopsLat tapOutStopsLon  \
0       P00253                Tegalan      -6.203101      106.85715   
1      B03307P    Sampoerna Strategic      -6.217152      106.81892   
2      B04962P  Simpang Kunir Kemukus      -6.133731      106.81475   

   stopEndSeq           tapOutTime  payAmount  
0        12.0  2023-04-03 06:00:53     3500.0  
1        21.0  2023-04-03 06:40:01     3500.0  
2        39.0  2023-04-03 06:50:55     3500.0  

[3 rows x 22 columns]

transID        payCardID payCardBank             payCardName  \
37897  YXPP627N4G95HO  213159426675861      emoney  drg. Zahra Nashiruddin   
37898  RGVK175U2U98UV  377840859133591      emoney            Ana Agustina   
37899  FMZZ963S4B68ZP     501862539795       flazz       drg. Leo Najmudin   

      payCardSex  payCardBirthDate corridorID                 corridorName  \
37897          F              1976         1T         Cibubur - Balai Kota   
37898          M              1976     JAK.13  Tanah Abang - Jembatan Lima   
37899          F              1985         13            Ciledug - Tendean   

       direction tapInStops  ... tapInStopsLon  stopStartSeq  \
37897        1.0    B02873P  ...     106.81676            12   
37898        1.0    B02505P  ...     106.80954            33   
37899        0.0     P00001  ...     106.74786             2   

                 tapInTime  tapOutStops  tapOutStopsName tapOutStopsLat  \
37897  2023-04-04 10:29:47      B00226P  Buperta Cibubur      -6.370321   
37898  2023-04-15 19:59:26      B01787P       JPO Blok G      -6.188861   
37899  2023-04-12 21:08:12       P00106   Kebayoran Lama      -6.238340   

      tapOutStopsLon  stopEndSeq           tapOutTime  payAmount  
37897      106.89628        14.0  2023-04-04 13:27:25    20000.0  
37898      106.81135        34.0  2023-04-15 20:27:50        0.0  
37899      106.77752         7.0  2023-04-12 21:34:53     3500.0  

[3 rows x 22 columns]

### **Data Understanding**

In [3]:
# Checking data types and info
print("Data Types:", df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37900 entries, 0 to 37899
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   transID           37900 non-null  object 
 1   payCardID         37900 non-null  int64  
 2   payCardBank       37900 non-null  object 
 3   payCardName       37900 non-null  object 
 4   payCardSex        37900 non-null  object 
 5   payCardBirthDate  37900 non-null  int64  
 6   corridorID        36643 non-null  object 
 7   corridorName      35970 non-null  object 
 8   direction         37900 non-null  float64
 9   tapInStops        36687 non-null  object 
 10  tapInStopsName    37900 non-null  object 
 11  tapInStopsLat     37900 non-null  float64
 12  tapInStopsLon     37900 non-null  float64
 13  stopStartSeq      37900 non-null  int64  
 14  tapInTime         37900 non-null  object 
 15  tapOutStops       35611 non-null  object 
 16  tapOutStopsName   36556 non-null  object

In [4]:
display(df.describe(), df.describe(include='object'))

payCardID  payCardBirthDate     direction  tapInStopsLat  \
count  3.790000e+04      37900.000000  37900.000000   37900.000000   
mean   4.250060e+17       1990.089314      0.500633      -6.214838   
std    1.321699e+18         13.051482      0.500006       0.057911   
min    6.040368e+10       1946.000000      0.000000      -6.394973   
25%    1.800442e+14       1982.000000      0.000000      -6.245863   
50%    3.507947e+15       1990.000000      1.000000      -6.214587   
75%    4.699023e+15       2001.000000      1.000000      -6.175528   
max    4.997694e+18       2012.000000      1.000000      -6.089429   

       tapInStopsLon  stopStartSeq  tapOutStopsLat  tapOutStopsLon  \
count   37900.000000  37900.000000    36556.000000    36556.000000   
mean      106.841554     13.572480       -6.214651      106.841233   
std         0.060369     12.237623        0.059022        0.060999   
min       106.614730      0.000000       -6.394973      106.614730   
25%       106.803470      4.000000       -6.247225      106.801750   
50%       106.834830     10.000000       -6.214718      106.834580   
75%       106.882270     19.000000       -6.174736      106.883030   
max       107.023950     68.000000       -6.091746      107.023660   

         stopEndSeq     payAmount  
count  36556.000000  36893.000000  
mean      21.219909   2699.712683  
std       13.800689   4212.225592  
min        1.000000      0.000000  
25%       11.000000      0.000000  
50%       18.000000   3500.000000  
75%       29.000000   3500.000000  
max       77.000000  20000.000000

transID payCardBank  payCardName payCardSex corridorID  \
count            37900       37900        37900      37900      36643   
unique           37900           6         1993          2        221   
top     FMZZ963S4B68ZP         dki  Suci Wacana          F         1T   
freq                 1       18743           80      20157        400   

                corridorName tapInStops tapInStopsName            tapInTime  \
count                  35970      36687          37900                37900   
unique                   216       2570           2602                37079   
top     Cibubur - Balai Kota     P00170    Penjaringan  2023-04-06 17:35:40   
freq                     391        236            243                    3   

       tapOutStops tapOutStopsName           tapOutTime  
count        35611           36556                36556  
unique        2230            2248                35908  
top         P00016             BKN  2023-04-24 06:53:50  
freq           306             316                    3

In [5]:
# Data Unique pada tiap kolom
pd.set_option('display.max_colwidth',None)
listItem = []
for col in df.columns :
    listItem.append( [col, df[col].nunique(), df[col].unique()])

tabel1Desc = pd.DataFrame(columns=['Column Name', 'Number of Unique', 'Unique Sample'],
                     data=listItem)
tabel1Desc

Column Name  Number of Unique  \
0            transID             37900   
1          payCardID              2000   
2        payCardBank                 6   
3        payCardName              1993   
4         payCardSex                 2   
5   payCardBirthDate                67   
6         corridorID               221   
7       corridorName               216   
8          direction                 2   
9         tapInStops              2570   
10    tapInStopsName              2602   
11     tapInStopsLat              2587   
12     tapInStopsLon              2458   
13      stopStartSeq                67   
14         tapInTime             37079   
15       tapOutStops              2230   
16   tapOutStopsName              2248   
17    tapOutStopsLat              2237   
18    tapOutStopsLon              2139   
19        stopEndSeq                74   
20        tapOutTime             35908   
21         payAmount                 3   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           Unique Sample  
0                                                                                                                                                                                                                                                                                                      

In [6]:
# Checking for missing values
print("Missing Values Percentage:")
df.isna().sum()/df.shape[0] * 100

Missing Values Percentage:


transID             0.000000
payCardID           0.000000
payCardBank         0.000000
payCardName         0.000000
payCardSex          0.000000
payCardBirthDate    0.000000
corridorID          3.316623
corridorName        5.092348
direction           0.000000
tapInStops          3.200528
tapInStopsName      0.000000
tapInStopsLat       0.000000
tapInStopsLon       0.000000
stopStartSeq        0.000000
tapInTime           0.000000
tapOutStops         6.039578
tapOutStopsName     3.546174
tapOutStopsLat      3.546174
tapOutStopsLon      3.546174
stopEndSeq          3.546174
tapOutTime          3.546174
payAmount           2.656992
dtype: float64

In [7]:
# Checking for duplicate values
print("Count of duplicate value:")
df.duplicated().sum()

Count of duplicate value:


np.int64(0)

### **Data Cleaning**

**Step by step data cleaning**
1. Drop column yang tidak relevan untuk kebutuhan analysis
2. Checking missing value pada setiap column
3. Mengisi missing value dengan metode tertentu : 
- Mengisi value pada kolom dengan cara mengzip dengan kolom lain
- Mengisi missing dengan domain knowledge
- Drop raw yang terdapat missing value
4. Menambahkan kolom untuk mempermudah analisis
5. Melakukan normalisasi pada value di setiap column
6. Fixing data type

In [8]:
# Dropping duplicate values
df.drop_duplicates(inplace=True)

In [9]:
# Drop Columns 
df = df.drop(columns=['tapInStopsLat', 'tapInStopsLon', 'tapOutStopsLat', 'tapOutStopsLon', 'payCardName', 
                      'transID', 'payCardName', 'stopEndSeq', 'stopStartSeq'])
df.head(3)

payCardID payCardBank payCardSex  payCardBirthDate corridorID  \
0   180062659848800      emoney          M              2008          5   
1  4885331907664776         dki          F              1997         6C   
2  4996225095064169         dki          F              1992        R1A   

                               corridorName  direction tapInStops  \
0                     Matraman Baru - Ancol        1.0     P00142   
1  Stasiun Tebet - Karet via Patra Kuningan        0.0    B01963P   
2                        Pantai Maju - Kota        0.0    B00499P   

  tapInStopsName            tapInTime tapOutStops        tapOutStopsName  \
0      Pal Putih  2023-04-03 05:21:44      P00253                Tegalan   
1     Kemenkes 2  2023-04-03 05:42:44     B03307P    Sampoerna Strategic   
2   Gg. Kunir II  2023-04-03 05:59:06     B04962P  Simpang Kunir Kemukus   

            tapOutTime  payAmount  
0  2023-04-03 06:00:53     3500.0  
1  2023-04-03 06:40:01     3500.0  
2  2023-04-03 06:50:55     3500.0

In [10]:
df.isna().sum()/df.shape[0] * 100

payCardID           0.000000
payCardBank         0.000000
payCardSex          0.000000
payCardBirthDate    0.000000
corridorID          3.316623
corridorName        5.092348
direction           0.000000
tapInStops          3.200528
tapInStopsName      0.000000
tapInTime           0.000000
tapOutStops         6.039578
tapOutStopsName     3.546174
tapOutTime          3.546174
payAmount           2.656992
dtype: float64

In [11]:
# Fill the corridorID column by zipping with the corridorName column
corridor_map = dict(zip(df["corridorName"], df["corridorID"]))
df["corridorID"] = df["corridorName"].replace(corridor_map)

In [12]:
df[['corridorID']].isna().sum()

corridorID    150
dtype: int64

In [13]:
# Pengecekan kolom corridorID yang masih kosong
df.loc[df['corridorID'].isna()].head()

payCardID payCardBank payCardSex  payCardBirthDate corridorID  \
206  4166272280740025         dki          F              2012        NaN   
230  3587585995659591         dki          F              1982        NaN   
506  4166272280740025         dki          F              2012        NaN   
530  3587585995659591         dki          F              1982        NaN   
806  4166272280740025         dki          F              2012        NaN   

                                     corridorName  direction tapInStops  \
206  Rusun Marunda - Terminal Terpadu Pulo Gebang        1.0    B04368P   
230  Rusun Marunda - Terminal Terpadu Pulo Gebang        0.0    B05937P   
506  Rusun Marunda - Terminal Terpadu Pulo Gebang        0.0    B03827P   
530  Rusun Marunda - Terminal Terpadu Pulo Gebang        1.0    B04368P   
806  Rusun Marunda - Terminal Terpadu Pulo Gebang        1.0    B04368P   

                tapInStopsName            tapInTime tapOutStops  \
206        Sbr. SDN Rorotan 03  2023-04-03 05:30:08         NaN   
230               YPI Asyaadah  2023-04-03 06:56:55     B00415P   
506  Sbr. Jln. Rorotan V Gg. 1  2023-04-03 17:24:21     B04086P   
530        Sbr. SDN Rorotan 03  2023-04-03 16:40:01     B02222P   
806        Sbr. SDN Rorotan 03  2023-04-04 05:43:19     B04608P   

                 tapOutStopsName           tapOutTime  payAmount  
206                          NaN                  NaN        0.0  
230    Gg. Barokah Lingkar Timur  2023-04-03 07:38:30        0.0  
506      Sbr. Mesindo Tekninesia  2023-04-03 18:36:30        0.0  
530  Markas Intai Amfibi Marunda  2023-04-03 18:34:59        0.0  
806          Sbr. MIN 20 Jakarta  2023-04-04 06:17:50        0.0

In [14]:
df.loc[df['corridorID'].isna(),'corridorName'].unique()

array(['Rusun Marunda - Terminal Terpadu Pulo Gebang'], dtype=object)

kolom corridorID yang memiliki NaN hanya pada jurusan Rusun Marunda - Terminal Terpadu Pulo Gebang maka kita akan mengisinya dengan nilai JAK110A sesuai dengan referensi yang ada pada https://moovitapp.com/index/id/Tranportasi_Umum-line-JAK110A-Jakarta-2044-2026517-111162611-0

In [15]:
# Mengisi corridorID yang berisi jurusan Rusun Marunda - Terminal Terpadu Pulo Gebang
df.loc[df['corridorID'].isna(), 'corridorID'] = 'JAK110A'

In [16]:
df[['corridorID']].isna().sum()

corridorID    0
dtype: int64

In [17]:
# Pengecekan kolom corridorName yang masih kosong
df.loc[df['corridorName'].isna()].head()

payCardID payCardBank payCardSex  payCardBirthDate corridorID  \
7   3587341902618993         dki          F              1991         3B   
38  4947438345710982         dki          F              2006         3B   
39  3539535695881465         dki          M              1986         3B   
60     4290559479079      online          M              2011         3B   
61   213155263532652      emoney          F              1984         3B   

   corridorName  direction tapInStops      tapInStopsName  \
7           NaN        1.0    B03416P  SDN Pondok Labu 11   
38          NaN        1.0    B02994P        PTUN Jakarta   
39          NaN        0.0     P00202        Rawa Selatan   
60          NaN        1.0    B00589P         GBK Pintu 7   
61          NaN        0.0    B05324P             SMKN 56   

              tapInTime tapOutStops         tapOutStopsName  \
7   2023-04-03 06:44:24     B00899P              Jln. Baros   
38  2023-04-03 05:29:43     B05707P  Term. Kampung Melayu 2   
39  2023-04-03 05:28:22      P00137                   Monas   
60  2023-04-03 05:43:48     B02556P            Palem Semi 1   
61  2023-04-03 06:13:03      P00138       Museum Fatahillah   

             tapOutTime  payAmount  
7   2023-04-03 07:11:43        0.0  
38  2023-04-03 05:58:35     3500.0  
39  2023-04-03 06:02:36     3500.0  
60  2023-04-03 06:09:12    20000.0  
61  2023-04-03 06:50:23     3500.0

kolom corridorName yang memiliki NaN hanya ada dengan ID 3B maka kita akan mengisinya dengan nilai Flamboyan-Kalideres dengan referensi dari https://megapolitan.kompas.com/read/2023/01/12/00150001/rute-transjakarta-3b-rusun-flamboyan-kalideres.

In [18]:
# Mengisi missing value dari corridorName yang berisi ID 3B
df.loc[df['corridorName'].isna(), 'corridorName'] = 'Rusun Flamboyan-Kalideres'

In [19]:
# Pengecekkan ulang data kosong
df[['corridorName']].isna().sum()

corridorName    0
dtype: int64

Karena jumlah missing value yang terdapat pada kolom tapInStops dan hanya sebesar < 5% maka kita akan menghapus baris yang memiliki missing value pada kolom tapInStops

In [20]:
# Menghapus baris yang terdapat missing value dari tapInStops
df = df.loc[df['tapInStops'].notna()]

In [21]:
# Pengecekan baris yang terdapat nilai NaN pada kolom tapOutStopsName
df.loc[df['tapOutStopsName'].isnull()].head()

payCardID payCardBank payCardSex  payCardBirthDate corridorID  \
6     213155822653833      emoney          M              1974     JAK.18   
30   3589380857722689         dki          M              1995     JAK.46   
70   3549158005951908         dki          M              1981         1T   
130      570918242399       flazz          F              1998        10D   
192  3548650711273610         dki          F              2011         3F   

                         corridorName  direction tapInStops  \
6                 Kalibata - Kuningan        0.0    B00243P   
30           Pasar Minggu - Jagakarsa        1.0    B00001P   
70               Cibubur - Balai Kota        0.0    B02408P   
130  Kampung Rambutan - Tanjung Priok        0.0     P00035   
192     Kalideres - Gelora Bung Karno        0.0     P00050   

            tapInStopsName            tapInTime tapOutStops tapOutStopsName  \
6          Cervino Village  2023-04-03 06:58:21         NaN             NaN   
30          18 Office Park  2023-04-03 06:14:02         NaN             NaN   
70            Menara Astra  2023-04-03 05:57:55         NaN             NaN   
130            Cempaka Mas  2023-04-03 05:40:04         NaN             NaN   
192  Dispenda Samsat Barat  2023-04-03 06:15:53         NaN             NaN   

    tapOutTime  payAmount  
6          NaN        0.0  
30         NaN        0.0  
70         NaN    20000.0  
130        NaN     3500.0  
192        NaN     3500.0

Bisa kita lihat jika baris yang terdapat nilai NaN pada kolom tapOutStopName maka akan secara otomatis kolom tapOutTime akan berisikan nilai NaN juga,Karena jika user tidak melakukan tap out ketika turun dari bis maka kolom tapOutStops,tapOutTime dan tapOutStopName tidak akan terisi dan kita tidak dapat melakukan asumsi atau mencari referensi apabila baris yang terdapat nilai NaN pada kolom tersebut,oleh karena itu dapat kita drop baris yang memiliki nilai NaN pada 3 kolom tersebut

In [22]:
# Menghapus baris yang terdapat missing value pada kolom tapOutStops
idx = df[df['tapOutStops'].isnull()].index
df = df.drop(labels=idx)

Untuk mengisi nilai pada kolom payAmount kita dapat mengisinya dengan cara mencari referensi yang terdapat pada internet dan mengkategorikannya dengan menyesuaikan data apa saja yang kita miliki

In [23]:
# Mengecek jumlah pembayaran
df['payAmount'].unique()

array([ 3500., 20000.,     0.,    nan])

https://transjakarta.co.id/faq-transjakarta/

Dari ketentuan Transjakarta yang saya dapat pada website resmi milik Transjakarta tersebut dapat kita kategorikan dengan menyesuaikan data yang kita miliki dengan asumsi :

- Untuk mengisi missing value dengan range umur > 60 tahun maka kita akan mengisi nya dengan nilai 0 sesuai dengan ketentuan Transjakarta
- Untuk mengisi missing value dengan jurusan yang menggunakan Royal Transjakarta kita akan mengisinya dengan nilai 20000 sesuai dengan ketentuan Transjakarta
- Untuk sisanya kita akan mengisi nilai 3500 dengan asumsi penumpang dengan umur < 60 tahun dan tidak menggunakan jurusan dengan bis Royal Transjakarta

In [24]:
# Mengisi baris dengan missing value pada kolom payAmount yang sesuai dengan ketentuan
df.loc[(df['payCardBirthDate'] <= 1963) & (df['payAmount'].isna()), 'payAmount'] = 0.0

In [25]:
# Pengecekan penumpang yang memakai Royal Transjakarta dengan melihat pada kolom payAmout
df.loc[df['payAmount'] == 20000.0]['corridorID'].unique()

array(['1T', 'B14', '3B', 'T21', 'D32', 'S31', 'B13', 'D31', '1K', '6P',
       'S12'], dtype=object)

In [26]:
# Membuat list corridorID yang menggunakan Royal Transjakarta
corridor_list = ['1T', 'B14', '3B', 'T21', 'D32', 'S31', 'B13', 'D31', '1K', '6P', 'S12']

In [27]:
# Mengisi nilai missing value dengan asumsi yang menggunakan bis dengan jurusan Royal Transjakarta
df.loc[df['corridorID'].isin(corridor_list) & df['payAmount'].isna(), 'payAmount'] = 20000.0

In [28]:
# Pengecekan missing value pada kolom payAmount
df[['payAmount']].isna().sum()

payAmount    852
dtype: int64

Dapat kita lihat masih terdapat jumlah missing value 852 dengan asumsi penumpang berumur dibawah 60 tahun dan tidak menggunakan Royal Transjakarta,oleh karena itu dapat kita isi dengan nilai 3500

In [29]:
# Mengisi nilai missing value dengan asumsi seperti diatas
df.loc[df['payAmount'].isna(), 'payAmount'] = 3500.0

In [30]:
# Pengecekan missing value pada kolom payAmount
df[['payAmount']].isna().sum()

payAmount    0
dtype: int64

In [31]:
# Mengecek nilai maximum dari kolom tapInTime
df['tapInTime'].max()

'2023-04-30 21:51:08'

In [32]:
# Mengecek nilai minimum dari kolom tapInTime
df['tapInTime'].min()

'2023-04-01 06:22:27'

Kita dapat melihat data tersebut adalah data yang ada ketika tahun 2023 sehingga kita dapat mencari umur dengan cara mengurangi tahun lahir user dengan tahun 2023 sehingga dapat diperoleh data umur dari setiap user

In [33]:
# Membuat kolom baru Age berdasarkan kolom payCardBirthDate
df['Age'] =  2023 - df['payCardBirthDate']
df.insert(3, 'Age',df.pop('Age'))
df.head()

payCardID payCardBank payCardSex  Age  payCardBirthDate corridorID  \
0   180062659848800      emoney          M   15              2008          5   
1  4885331907664776         dki          F   26              1997         6C   
2  4996225095064169         dki          F   31              1992        R1A   
3      639099174703       flazz          F   45              1978        11D   
4      570928206772       flazz          M   41              1982         12   

                               corridorName  direction tapInStops  \
0                     Matraman Baru - Ancol        1.0     P00142   
1  Stasiun Tebet - Karet via Patra Kuningan        0.0    B01963P   
2                        Pantai Maju - Kota        0.0    B00499P   
3       Pulo Gebang - Pulo Gadung 2 via PIK        0.0    B05587P   
4                     Tanjung Priok - Pluit        0.0     P00239   

           tapInStopsName            tapInTime tapOutStops  \
0               Pal Putih  2023-04-03 05:21:44      P00253   
1              Kemenkes 2  2023-04-03 05:42:44     B03307P   
2            Gg. Kunir II  2023-04-03 05:59:06     B04962P   
3            Taman Elok 1  2023-04-03 05:44:51     B03090P   
4  Sunter Boulevard Barat  2023-04-03 06:17:35      P00098   

         tapOutStopsName           tapOutTime  payAmount  
0                Tegalan  2023-04-03 06:00:53     3500.0  
1    Sampoerna Strategic  2023-04-03 06:40:01     3500.0  
2  Simpang Kunir Kemukus  2023-04-03 06:50:55     3500.0  
3      Raya Penggilingan  2023-04-03 06:28:16     3500.0  
4       Kali Besar Barat  2023-04-03 06:57:03     3500.0

In [34]:
# Menghapus kolom payCardBirthDate yang berisi tahun kelahiran
df = df.drop(columns='payCardBirthDate')

In [35]:
# Mengubah tipe data kolom 'tapInTime' dan 'tapOutTime' menjadi datetime
df['tapInTime'] = pd.to_datetime(df['tapInTime'])
df['tapOutTime'] = pd.to_datetime(df['tapOutTime'])

In [36]:
# Mengecek type data kolom tapInTime dan tapOutTime
df[['tapInTime','tapOutTime']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 34460 entries, 0 to 37899
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   tapInTime   34460 non-null  datetime64[ns]
 1   tapOutTime  34460 non-null  datetime64[ns]
dtypes: datetime64[ns](2)
memory usage: 807.7 KB


In [37]:
# Membuat kolom tapInTime,tapInDay,dan tapInDate
df['tapInTime2'] = df['tapInTime'].dt.strftime('%H:%M')
df['tapInDay'] = df['tapInTime'].dt.strftime('%A')
df['tapInDate'] = df['tapInTime'].dt.strftime('%m-%d')
df.insert(8, 'tapInTime2',df.pop('tapInTime2'))
df.insert(9, 'tapInDay',df.pop('tapInDay'))
df.insert(10, 'tapInDate',df.pop('tapInDate'))

In [38]:
# Menghapus kolom tapInTime
df = df.drop(columns='tapInTime')

In [39]:
# Mengganti nama kolom tapInTime2 menjadi tapInTime
df = df.rename(columns={'tapInTime2' : 'tapInTime'})

In [40]:
# Membuat kolom tapOutTime,tapOutDay,dan tapOutDate
df['tapOutTime2'] = df['tapOutTime'].dt.strftime('%H:%M')
df['tapOutDay'] = df['tapOutTime'].dt.strftime('%A')
df['tapOutDate'] = df['tapOutTime'].dt.strftime('%m-%d')
df.insert(13, 'tapOutTime2',df.pop('tapOutTime2'))
df.insert(14, 'tapOutDay',df.pop('tapOutDay'))
df.insert(15, 'tapOutDate',df.pop('tapOutDate'))

In [41]:
# Menghapus kolom tapOutTime
df = df.drop(columns='tapOutTime')

In [42]:
# Mengganti nama tapOutTime2 menjadi tapOutTime
df = df.rename(columns={'tapOutTime2' : 'tapOutTime'})

### **Data yang sudah bersih**

In [43]:
df.isna().sum()

payCardID          0
payCardBank        0
payCardSex         0
Age                0
corridorID         0
corridorName       0
direction          0
tapInStops         0
tapInTime          0
tapInDay           0
tapInDate          0
tapInStopsName     0
tapOutStops        0
tapOutTime         0
tapOutDay          0
tapOutDate         0
tapOutStopsName    0
payAmount          0
dtype: int64

In [44]:
listItem = []
for col in df.columns :
    listItem.append([col, df[col].dtype, df[col].isna().sum(),
                    df[col].nunique(), list(df[col].drop_duplicates().sample(2).values)]);

dfDesc = pd.DataFrame(columns=['dataFeatures', 'dataType', 'null', 'unique', 'uniqueSample'],
                     data=listItem)
print(df.shape)
dfDesc

(34460, 18)


dataFeatures dataType  null  unique  \
0         payCardID    int64     0    1971   
1       payCardBank   object     0       6   
2        payCardSex   object     0       2   
3               Age    int64     0      66   
4        corridorID   object     0     216   
5      corridorName   object     0     217   
6         direction  float64     0       2   
7        tapInStops   object     0    2508   
8         tapInTime   object     0    1014   
9          tapInDay   object     0       7   
10        tapInDate   object     0      30   
11   tapInStopsName   object     0    2507   
12      tapOutStops   object     0    2198   
13       tapOutTime   object     0    1116   
14        tapOutDay   object     0       7   
15       tapOutDate   object     0      30   
16  tapOutStopsName   object     0    2197   
17        payAmount  float64     0       3   

                                                            uniqueSample  
0                                   [4317815230091374, 6011309044185857]  
1                                                          [online, dki]  
2                                                                 [F, M]  
3                                                               [46, 77]  
4                                                          [11C, JAK.50]  
5   [Ragunan - MH Thamrin via Semanggi, Kelapa Gading - Rusun Kemayoran]  
6                                                             [1.0, 0.0]  
7                                                     [B02542P, B00829P]  
8                                                         [17:04, 14:12]  
9                                                     [Saturday, Sunday]  
10                                                        [04-13, 04-27]  
11     [Sbr. Jln. Bendungan Hilir Gg. IV, Sbr. Jln. Flamboyan Srengseng]  
12                                                    [B03094P, B03240P]  
13                                                        [14:29, 11:35]  
14                                                    [Friday, Saturday]  
15                                                        [04-07, 04-01]  
16                      [Term. Kampung Melayu 2, Jln. Danau Indah Timur]  
17                                                     [3500.0, 20000.0]

In [45]:
df.head()

payCardID payCardBank payCardSex  Age corridorID  \
0   180062659848800      emoney          M   15          5   
1  4885331907664776         dki          F   26         6C   
2  4996225095064169         dki          F   31        R1A   
3      639099174703       flazz          F   45        11D   
4      570928206772       flazz          M   41         12   

                               corridorName  direction tapInStops tapInTime  \
0                     Matraman Baru - Ancol        1.0     P00142     05:21   
1  Stasiun Tebet - Karet via Patra Kuningan        0.0    B01963P     05:42   
2                        Pantai Maju - Kota        0.0    B00499P     05:59   
3       Pulo Gebang - Pulo Gadung 2 via PIK        0.0    B05587P     05:44   
4                     Tanjung Priok - Pluit        0.0     P00239     06:17   

  tapInDay tapInDate          tapInStopsName tapOutStops tapOutTime tapOutDay  \
0   Monday     04-03               Pal Putih      P00253      06:00    Monday   
1   Monday     04-03              Kemenkes 2     B03307P      06:40    Monday   
2   Monday     04-03            Gg. Kunir II     B04962P      06:50    Monday   
3   Monday     04-03            Taman Elok 1     B03090P      06:28    Monday   
4   Monday     04-03  Sunter Boulevard Barat      P00098      06:57    Monday   

  tapOutDate        tapOutStopsName  payAmount  
0      04-03                Tegalan     3500.0  
1      04-03    Sampoerna Strategic     3500.0  
2      04-03  Simpang Kunir Kemukus     3500.0  
3      04-03      Raya Penggilingan     3500.0  
4      04-03       Kali Besar Barat     3500.0

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 34460 entries, 0 to 37899
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   payCardID        34460 non-null  int64  
 1   payCardBank      34460 non-null  object 
 2   payCardSex       34460 non-null  object 
 3   Age              34460 non-null  int64  
 4   corridorID       34460 non-null  object 
 5   corridorName     34460 non-null  object 
 6   direction        34460 non-null  float64
 7   tapInStops       34460 non-null  object 
 8   tapInTime        34460 non-null  object 
 9   tapInDay         34460 non-null  object 
 10  tapInDate        34460 non-null  object 
 11  tapInStopsName   34460 non-null  object 
 12  tapOutStops      34460 non-null  object 
 13  tapOutTime       34460 non-null  object 
 14  tapOutDay        34460 non-null  object 
 15  tapOutDate       34460 non-null  object 
 16  tapOutStopsName  34460 non-null  object 
 17  payAmount        

In [47]:
# Save df to xlsx
# df.to_excel('Transjakarta_cleaned.xlsx', index=False)